### PR TITLE
Simple normalisation of CompositeKeys

### PR DESCRIPTION
**Pull request by kasiastreich**
There was an issue with semantically equivalent trees but different structure, which causes equality and hashCode bugs. Added simple invariants to CompositeKey constructor to forbid the most obvious cases of such behaviour. Adjusted builder to emit simplified keys - if CompositeKey holds only single node, that node is moved up, so we won't end up with chains of wrapped keys. Added ordering of nodes on given tree level. It's not full normalisation of trees, but I think that it's sufficient (we don't have a use case for logical transformation of trees to the canonical form).

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/crypto/CryptoUtilities.kt [hunk 1]
```diff
@@ -115,20 +113,20 @@ fun PublicKey.toStringShort(): String {
...
 }
 
+fun PublicKey.isFulfilledBy(key: PublicKey): Boolean  = isFulfilledBy(setOf(key))
+fun PublicKey.isFulfilledBy(keys: Iterable<PublicKey>): Boolean {
+    return if (this is CompositeKey) this.isFulfilledBy(keys)
+    else this in keys
+}
+
+/** Checks whether any of the given [keys] matches a leaf on the CompositeKey tree or a single PublicKey */
+fun PublicKey.containsAny(otherKeys: Iterable<PublicKey>) = keys.intersect(otherKeys).isNotEmpty()
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CryptoUtilities.kt]: Could you expand this out so it doesn't unnecessarily box an individual key in a set just for this test? I know it means some code repetition, but I think makes it a bit easier to understand exactly what's going on.

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
 
 import net.corda.core.serialization.CordaSerializable
-import net.corda.core.serialization.deserialize
 import net.corda.core.serialization.serialize
-import java.security.InvalidKeyException
 import java.security.PublicKey
 
+// Holds node - weight pairs for a CompositeKey. Ordered first by weight, then by node's hashCode.
+@CordaSerializable
+data class NodeWeight(val node: PublicKey, val weight: Int): Comparable<NodeWeight> {
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Might be better to put this inside `CompositeKey`. It's not going to be used outside of the builder, or shouldn't be, should it? Could probably make it inaccessible from outside `CompositeKey` in fact.

File: core/src/main/kotlin/net/corda/core/serialization/Kryo.kt [hunk 3]
```diff
@@ -374,18 +375,16 @@ object CompositeKeySerializer : Serializer<CompositeKey>() {
...
-        output.writeInts(obj.weights.toIntArray())
     }
 
     override fun read(kryo: Kryo, input: Input, type: Class<CompositeKey>): CompositeKey {
         val threshold = input.readInt()
         val childCount = input.readInt()
-        val children = (1..childCount).map { kryo.readClassAndObject(input) as PublicKey }
-        val weights = input.readInts(childCount)
-
+        if (childCount <= 1) throw IllegalArgumentException("Trying to construct CompositeKey with duplicated child nodes.")
```
Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: Yes, message is wrong. But as to builder check, should we deserialize something that we already know is wrong?

Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: Oh, I see, it's validating the stored values could make sense before we try reading them. Can you replace it with:

`require (childCount > 1)`

Actually, do we have a utility function for "Read a list length then the objects that are part of the list" yet? If not we should add one

Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/serialization/Kryo.kt]: Now this pattern is used only in one place, isn't it?

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt (hunk 2 above)
Comment (kasiastreich) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Depends how you want to be able to construct CompositeKeys. I was also thinking about that, maybe we can make CompositeKey constructor private and allow only Builder?

File: finance/src/main/kotlin/net/corda/contracts/testing/DummyLinearContract.kt [hunk 4]
```diff
@@ -24,7 +24,7 @@ class DummyLinearContract: Contract {
             val nonce: SecureHash = SecureHash.randomSHA256()) : LinearState {
 
         override fun isRelevant(ourKeys: Set<PublicKey>): Boolean {
-            return participants.any { (it is CompositeKey && it.containsAny(ourKeys)) || it in ourKeys }
+            return participants.any { it.containsAny(ourKeys) }
```
Comment (kasiastreich) [inline finance/src/main/kotlin/net/corda/contracts/testing/DummyLinearContract.kt]: Yes, there is no default and it's repeated over and over again in different places of our code.

Comment (kasiastreich) [inline finance/src/main/kotlin/net/corda/contracts/testing/DummyLinearContract.kt]: It will be a separate PR.

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 5]
```diff
@@ -18,16 +16,32 @@ import java.security.PublicKey
  * Using these constructs we can express e.g. 1 of N (OR) or N of N (AND) signature requirements. By nesting we can
  * create multi-level requirements such as *"either the CEO or 3 of 5 of his assistants need to sign"*.
  *
- * [CompositeKey] maintains a list of child nodes – sub-trees, and associated
- * [weights] carried by child node signatures.
+ * [CompositeKey] maintains a list of [NodeWeigh] which holds child subtree with associated weight carried by child node signatures.
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: typo NodeWeight

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 6]
```diff
@@ -18,16 +16,32 @@ import java.security.PublicKey
...
+    val children = children.sorted()
+    init {
+        require (children.size == children.toSet().size) { "Trying to construct CompositeKey with duplicated child nodes." }
+        // If we want PublicKey we only keep one key, otherwise it will lead to semantically equivalent trees but having different structures.
+        require(children.size > 1) { "Cannot construct CompositeKey with only one child node." }
+    }
+
+    // Holds node - weight pairs for a CompositeKey. Ordered first by weight, then by node's hashCode.
+    @CordaSerializable
+    data class NodeWeight(val node: PublicKey, val weight: Int): Comparable<NodeWeight> {
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Maybe call it `NodeAndWeight`?

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 7]
```diff
@@ -18,16 +16,32 @@ import java.security.PublicKey
...
+class CompositeKey private constructor (val threshold: Int,
+                   children: List<NodeWeight>) : PublicKey {
+    val children = children.sorted()
+    init {
+        require (children.size == children.toSet().size) { "Trying to construct CompositeKey with duplicated child nodes." }
+        // If we want PublicKey we only keep one key, otherwise it will lead to semantically equivalent trees but having different structures.
+        require(children.size > 1) { "Cannot construct CompositeKey with only one child node." }
+    }
+
+    // Holds node - weight pairs for a CompositeKey. Ordered first by weight, then by node's hashCode.
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: use api comment block /**

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 8]
```diff
@@ -18,16 +16,32 @@ import java.security.PublicKey
...
+    @CordaSerializable
+    data class NodeWeight(val node: PublicKey, val weight: Int): Comparable<NodeWeight> {
+        override fun compareTo(other: NodeWeight): Int {
+            if(weight == other.weight) {
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: missing space after if

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 9]
```diff
@@ -37,24 +51,21 @@ class CompositeKey(val threshold: Int,
...
     override fun getEncoded(): ByteArray = this.serialize().bytes
     override fun getFormat() = FORMAT
 
     // TODO Can CompositeKey be fulfilled by other composite keys? With composite signature it makes some sense.
     fun isFulfilledBy(keys: Iterable<PublicKey>): Boolean {
-        val totalWeight = children.mapIndexed { i, childNode ->
-            if (childNode is CompositeKey) {
-                if (childNode.isFulfilledBy(keys))
-                    weights[i]
+        val totalWeight = children.map { it ->
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: could use Kotlin 1.1 destructuring and write `(node, weight) ->` instead if `it`

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 10]
```diff
@@ -109,22 +112,23 @@ class CompositeKey(val threshold: Int,
...
          */
-        fun build(threshold: Int? = null): CompositeKey {
-            return CompositeKey(threshold ?: children.size, children.toList(), weights.toList())
+        @Throws(IllegalArgumentException::class)
+        fun build(threshold: Int? = null): PublicKey {
+            val n = children.size
+            if (n > 1)
+                return CompositeKey(threshold ?: n, children)
+            else if (n == 1) {
+                if (threshold != null && threshold != children[0].weight)
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: can use `require()`

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 11]
```diff
@@ -109,22 +112,23 @@ class CompositeKey(val threshold: Int,
...
+                    throw IllegalArgumentException("Trying to build invalid CompositeKey, threshold value different than weight of single child node.")
+                return children[0].node // We can assume that this node is a correct CompositeKey.
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: `children.first()` might be more readable

**Comment by kasiastreich** [comment]
Addressed all comments.
